src/main/java/org/broadinstitute/hellbender/utils/config/ConfigUtils.java

```diff
@@ -0,0 +1,332 @@
+
+            // Get the classes from which we need to look for sources:
+            Class<?>[] configurationClasses = new Class<?>[] {
+                    GATKConfig.class
+            };
+
+            // Create a regex to use to look for variables in the Sources annotation:
+            Pattern p = Pattern.compile("\\$\\{(.*)}");
+
+            // Loop through our classes and grab any sources with variables in there:
+            for ( Class<?> clazz : configurationClasses ) {
+                Set<Class<?>> interfaces = new HashSet<>(Arrays.asList(clazz.getInterfaces()));
+
+                // Make sure that we get config classes here:
+                if ( interfaces.contains(Config.class) ||
+                         interfaces.contains(Accessible.class) ||
+                         interfaces.contains(Mutable.class) ||
+                         interfaces.contains(Reloadable.class) ) {
+                    Config.Sources annotation = clazz.getAnnotation(Config.Sources.class);
+
+                    String[] annotationValues = annotation.value();
+
+                    for ( String val : annotationValues ) {
+
+                        Matcher m = p.matcher(val);
+                        if ( m.find() ) {
+                            propertyNames.add(m.group(1));
+                        }
+                    }
+                }
+            }
+
+            // Grab the system properties:
+            Properties systemProperties = System.getProperties();
+
+            // Grab the environment properties:
+            Map<String, String> environmentProperties = System.getenv();
+
+            // Make sure that if our property isn't in the system and environment
+            // properties, that we set it to a neutral value that will not contain
+            // anything (so that the property will fall back into the next value).
+            for (String property : propertyNames) {
+
+                if ((!environmentProperties.keySet().contains(property)) &&
+                        (!systemProperties.containsKey(property))) {
+
+                    ConfigFactory.setProperty(property, "/dev/null");
+                }
+            }
+
+            hasSetConfigFactoryVariableDefaults = true;
+        }
+    }
+
+    /**
+     * Get the configuration file name from the given arguments.
+     * Modifies the given arguments to remove both the configuration file specification string
+     * and the configuration file name from the args.
+     *
+     * NOTE: Does NOT validate that the resulting string is a valid configuration file.
+     *
+     * @param args Command-line arguments passed to this program.
+     * @param configFileOption The command-line option indicating that the config file is next
+     * @return The name of the configuration file for this program or {@code null}.
+     */
+    public static final String getConfigFilenameFromArgs( final ArrayList<String> args, final String configFileOption ) {
```

Comment:
I kind of like the idea of breaking this into two methods, one that gets the  value from the args, and a separate one that removes the argument.  I hate having these methods that mix return values with side effects.  At a minimum maybe give it a more descriptive name so it's clear what the side effects are.  It's ok if it's grossly long...

As a different question... is there a way we can get this argument to play nicely with barclay so we don't have to strip it out.  It's weird that there's going to be an arg that doesn't show up in the args list.  Can we make it one of the SpecialArguments and just pass it through and have barclay black hole it?